src/stream/src/common/log_store_impl/kv_log_store/serde.rs

```diff
@@ -654,6 +657,95 @@ impl<S: StateStoreReadIter> LogStoreRowOpStream<S> {
             }
         }
     }
+
+    #[expect(dead_code)]
+    #[try_stream(ok = (LogStoreVnodeProgress, KvLogStoreItem), error = anyhow::Error)]
+    async fn into_vnode_log_store_item_stream(mut self, chunk_size: usize) {
+        assert!(chunk_size >= 2, "too small chunk_size: {}", chunk_size);
+        let mut ops = Vec::with_capacity(chunk_size);
+        let mut data_chunk_builder =
+            DataChunkBuilder::new(self.serde.payload_schema.clone(), chunk_size);
+
+        let mut progress = LogStoreVnodeProgress::new();
```

Comment:
I don't get the semantic of `progress` map here. It seems that we take the progress every time we want to yield either a stream chunk or barrier, but for each item, it should only belong to a single epoch, so why store it as a map with epoch as key? Besides, in current implementation, since the key is the epoch, rows in the same epoch will overwrite the previous written progress, even if the progress is of the other vnode. Is this expected?

Reply:
Fixed https://github.com/risingwavelabs/risingwave/pull/21241/commits/d4cac585b350c735026699f894d0fbd4386dffd1